### PR TITLE
docs: add upgrade and migration section

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+To better align with the newly introduced support for testing justifications and indictments,
+the assertion methods which accepted a `message` argument now have it as the first argument.
+
+Before in `*ConstraintProviderTest.java`:
+
+[source,java]
+----
+constraintVerifier.verifyThat(MyConstraintProvider::myConstraint)
+    .given()
+    .penalizesBy(0, "There should no penalties");
```
Comment (triceo) [inline docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:194]:
```suggestion
    .penalizesBy(0, "There should be no penalties");
```

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+----
+
+After in `*ConstraintProvider.java`:
+
+[source,java]
+----
+constraintVerifier.verifyThat(MyConstraintProvider::myConstraint)
+    .given()
+    .penalizesBy("There should no penalties", 0);
```
Comment (triceo) [inline docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:203]:
```suggestion
    .penalizesBy("There should be no penalties", 0);
```

File: docs/src/modules/ROOT/nav.adoc [hunk 3]
```diff
@@ -4,6 +4,7 @@
 ** xref:quickstart/hello-world/hello-world-quickstart.adoc[leveloffset=+1]
 ** xref:quickstart/quarkus/quarkus-quickstart.adoc[leveloffset=+1]
 ** xref:quickstart/spring-boot/spring-boot-quickstart.adoc[leveloffset=+1]
+** xref:quickstart/quarkus-vehicle-routing/quarkus-vehicle-routing-quickstart.adoc[leveloffset=+1]
```
Comment (zepfred) [inline docs/src/modules/ROOT/nav.adoc:7]: Is this expected?

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc [hunk 4]
```diff
@@ -0,0 +1,213 @@
+[#upgradeToLatestVersion]
+= Upgrade to the latest version
+:doctype: book
+:sectnums:
+:icons: font
+
+Timefold Solver public APIs are backwards compatible,
+but users often also use impl classes,
+many of which are documented in the reference manual too.
```
Comment (zepfred) [inline docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:9]: Do we have a link for that?

File: docs/src/modules/ROOT/pages/upgrade-and-migration/overview.adoc [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+:icons: font
+
+In this section, you will:
+
+- Learn how to upgrade from OptaPlanner to Timefold Solver.
+- Learn how to upgrade to the latest version of Timefold Solver automatically.
+- Find a detailed upgrade recipe for every major change we make to the Timefold Solver APIs.
+
+To find out what's new in the latest version of Timefold Solver,
+see the [Timefold Solver release notes](https://github.com/TimefoldAI/timefold-solver/releases)
```
Comment (zepfred) [inline docs/src/modules/ROOT/pages/upgrade-and-migration/overview.adoc:14]: The 
```suggestion
see the https://github.com/TimefoldAI/timefold-solver/releases[Timefold Solver release notes]
```

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc [hunk 6]
```diff
@@ -0,0 +1,213 @@
+[#upgradeToLatestVersion]
+= Upgrade to the latest version
+:doctype: book
+:sectnums:
+:icons: font
+
+Timefold Solver public APIs are backwards compatible,
+but users often also use impl classes,
```
Comment (zepfred) [inline docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:8]: implementation classes?

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+[#manualUpgrade]
+== Manual upgrade recipe
+
+Every upgrade note indicates how likely your code will be affected by that change:
+
+- icon:magic[] *Automated*: Can be applied automatically using our <<automaticUpgradeToLatestVersion,migration tooling>>.
+- icon:exclamation-triangle[role=red] *Major*: Likely to affect your code.
+- icon:info-circle[role=yellow] *Minor*: Unlikely to affect your code, unless you use internal classes.
+- icon:info[] *Impl detail*: Will not affect your code, unless you are using some very dark corner of Solver's internals.
```
Comment (zepfred) [inline docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:57]: There's something off about this icon. What do you think?
Screenshot from 2024-03-08 10-49-39

**Review by zepfred — APPROVED** [review]
LGTM!

I have some minor suggestions. Also, the subsection `Upgrade to the latest version` under the `Introduction` section seems unnecessary now.

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc (hunk 7 above)
Comment (triceo) [reply docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:57]: I'm just going to remove this impl detail. It's not yet used - if it ever is, I'll figure it out then.

File: docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc (hunk 4 above)
Comment (triceo) [reply docs/src/modules/ROOT/pages/upgrade-and-migration/upgrade-to-latest-version.adoc:9]: There are many things like that. Move selectors, especially. Anyway, I'll rephrase it.

**Pull request merged**